update Clusters From Scratch for Fedora 21

kgaillot opened the pull request:
This revision is an extensive overhaul of the Clusters From Scratch documentation. Diff is possible but you'll probably want to compare the generated HTML side-by-side instead or in addition. For convenience I've generated the full branded docs and made them available at http://www.boogieonline.com/ha/ for comparison.

== The changes fall into these areas:
- Update for Fedora 21 (OS installation instructions, some systemd/firewalld-related stuff and pcs syntax/output).
- This should close http://bugs.clusterlabs.org/show_bug.cgi?id=5227 "typos in Pacemaker-1.1-Clusters_from_Scratch-en-US.pdf" and http://bugs.clusterlabs.org/show_bug.cgi?id=5228 "Clusters_from_Scratch 2.5.3. wrong command". The reporter's syntax suggestions may apply to F20 (I didn't check), but they definitely are not needed for F21. Only one of the recommended changes was needed (correcting the IP address in an example).
- Strip out the ifdef's for generating a crmsh version (internal markup changes only, no visible impact on generated output). I had mixed feelings about this, but the crmsh info would no longer have been valid as-is for Fedora 21, so I thought it was better to remove it altogether. (Version control makes me brave.)
- Make the styling more consistent with what the "Typographic Conventions" section says it is. This includes getting rid of the [source,C] highlighting for command-line blocks, since it's not really correct and the conventions do not indicate any highlighting in command-line output blocks.
- I added one stylistic convention, commands are displayed with '#' if they need to be done on both nodes, or '[root@pcmk-1 ~]#' if they need to be done on one specific node. I felt this was more consistent with what the reader would see if they were following exactly, but if you feel it clutters too much, we can go back to '#' everywhere.
- Make most headings imperative (e.g. "Adding a Resource" becomes "Add a Resource"). Psychological engineering to make the reader feel more active and confident.
- Move the STONITH chapter before Active/Active, because DLM now refuses to run if STONITH is disabled. Within Active/Active, I moved configuring a DLM resource before configuring GFS2, for clarity; before, the DLM configuration was in the middle of the GFS2 configuration.
- Misc. corrections. For example, the previous version had no colocation or ordering constraints for DLM and GFS2, so trying to stop cluster services on a node could lead to ordering problems and the node being fenced; those constraints were added.
- Misc. rewordings for clarity.

== Issues remaining:
- While I took out the crmsh ifdef's from the documentation, the Makefile.am in doc still distinguishes between crmsh and pcs versions. No harm done, but crmsh could be removed here, too -- but that would likely involve path changes (and thus .gitignore changes) etc. It would also be nice to redo some of the targets, e.g. allowing the full set of documentation to be built without also trying to upload it to clusterlabs.org.
- "2.6.2. Configure Corosync" has an ugly workaround for a bug in the currently shipped version of pcs in Fedora 21. Chris Feist is working on a fix, but in the meantime the only workarounds are disabling IPv6 on the node entirely or the one-line source change described in this section. I'm not sure whether we should hold CFS until the fix is available, or release it with the workaround and update it when the fix comes out.
- The whole process might not actually work if followed exactly; I encountered many issues and fixed or worked around them as I went, and I may have missed updating the documentation accordingly in some cases. It would be worthwhile for someone else to try to follow it to get a fresh perspective. One particularly important example: the cluster will sometimes put both ClusterIP clones on a single node, even if the other node is healthy. I couldn't quickly figure out constraints/attributes to alter that behavior, so I just didn't discuss it. I still need to figure out why that was happening and update the documentation once I know what to do about it.
- I didn't want to go down the rabbit hole of even bigger changes, but I think it would make sense to stop Chapter 2 at "Configure SSH" so that it was just the OS install/configuration, and put the remainder of chapter 2 with the current chapter 3 so that chapter 3 becomes cluster install/configuration.
- Finally and trivially, there are three places where this document is versioned. Book_Info.xml has edition and pubsnumber tags; the value of edition appears on the title page, and pubsnumber seems not be user-visible, so it's probably just for internal bookkeeping. Then Revision_History.xml has its own numbering system (and is visible as the final appendix). It looks weird that the title page has one version number and the revision history has another. I'd recommend keeping the Revision_History.xml X-Y versioning in sync with the EDITION-PUBSNUMBER of Book_info.xml (which would mean jumping EDITION ahead a few numbers). To put some logic behind it, we could increment EDITION whenever the version/syntax of the documented software changes (Fedora, pacemaker, corosync, pcs), and PUBSNUMBER for other changes (corrections, translations, etc.). But that's incredibly unimportant.

beekhof commented:
Splitting these up into smaller commits (particularly dropping the crmsh bits) would have helped readability, but this refresh was long overdue. Thanks!

beekhof commented on doc/Clusters_from_Scratch/en-US/Ch-Active-Active.txt:
```diff
@@ -1,755 +1,364 @@
+host the IP, so that if any node goes down, another node can take over
+the failed node's "request bucket". Otherwise, requests intended for
+the failed node would be discarded.
+* `globally-unique=true` tells the cluster that one clone isn't identical
```
globally-unique is also the trigger that turns on the hashing of packets so that only one node processes each request (even though both have the IP active)

kgaillot replied:
Will clarify globally-unique in next round of changes

beekhof commented on doc/Clusters_from_Scratch/en-US/Ch-Active-Active.txt:
```diff
@@ -1,755 +1,364 @@
+On both nodes, install the GFS2 command-line utilities and the
+Distributed Lock Manager (DLM) required by cluster filesystems:
+----
 # yum install -y gfs2-utils dlm kernel-modules-extra
```
At least one of dlm or gfs2-utils sucks it in.

kgaillot replied:
% rpm -q --whatrequires kernel-modules-extra
dlm-4.0.1-4.fc21.x86_64

will remove explicit install in next round

beekhof commented on doc/Clusters_from_Scratch/en-US/Ch-Apache.txt:
```diff
@@ -328,258 +175,143 @@ anywhere.
 -----
-# pcs constraint colocation add WebSite ClusterIP INFINITY
-# pcs constraint
+[root@pcmk-1 ~]# pcs constraint colocation add WebSite ClusterIP INFINITY
```
I like the 'with', it makes the whole thing much clearer

kgaillot replied:
Verified that 'with' works in F21, will add to next round of changes.

beekhof commented:
> Removing crmsh ifdef's from Makefile.am

go for it

> "2.6.2. Configure Corosync"  

Release it with the workaround and update it when the fix comes out.
Actually until everyone has the fix it is worth leaving the work-around as a note anyway.

> One particularly important example: the cluster will sometimes put both ClusterIP clones on a single > node, even if the other node is healthy.

This is certainly something worth highlighting and eventually fixing.
The problem is that clones are slightly sticky by default and don't rebalance after a failure.

The work-around is to force the stickiness of the clone to 0 temporarily.

> I didn't want to go down the rabbit hole of even bigger changes, but I think it would make sense to stop Chapter 2 at "Configure SSH" so that it was just the OS install/configuration, and put the remainder of chapter 2 with the current chapter 3 so that chapter 3 becomes cluster install/configuration.

sure

> To put some logic behind it, we could increment EDITION whenever the version/syntax of the documented software changes (Fedora, pacemaker, corosync, pcs), and PUBSNUMBER for other changes (corrections, translations, etc.).

Sounds good to me :)

kgaillot commented:
OK, I've added a second commit with these changes:
* Updated document versioning as described above
* Removed explicit install of kernel-modules-extra
* Clarified description of IPaddr2 attributes
* Added note about clone resource-stickiness=0 (I could swear I tried that before, but it worked when I tried it just now)
* Added 'with' to all colocation constraints

Re: smaller commits, I came to the same conclusion after it was too late to easily do :-)

Re: Makefile.am and chapter 2/3 reorg, I might do that when the pacemaker fix comes through, but I'll omit that for now.

I didn't regenerate the branded docs because the diff is readable this time. I did proof the changes in a local single-html copy.
